src/coreclr/vm/siginfo.cpp

```diff
@@ -5378,42 +5378,105 @@ MetaSig::TryGetUnmanagedCallingConventionFromModOpt(
+            return COR_E_INVALIDPROGRAM;
         }
+    }
 
-        const struct {
-            LPCSTR name;
-            CorInfoCallConvExtension value;
-        } knownCallConvs[] = {
-            { CMOD_CALLCONV_NAME_CDECL,     CorInfoCallConvExtension::C },
-            { CMOD_CALLCONV_NAME_STDCALL,   CorInfoCallConvExtension::Stdcall },
-            { CMOD_CALLCONV_NAME_THISCALL,  CorInfoCallConvExtension::Thiscall },
-            { CMOD_CALLCONV_NAME_FASTCALL,  CorInfoCallConvExtension::Fastcall } };
+    *suppressGCTransitionOut = ((modifiers & CALL_CONV_MOD_SUPPRESSGCTRANSITION) != 0);
 
-        for (const auto &callConv : knownCallConvs)
+    if (modifiers & CALL_CONV_MOD_MEMBERFUNCTION)
+    {
+        if (*callConvOut == CorInfoCallConvExtension::Managed)
         {
-            // Look for a recognized calling convention in metadata.
-            if (::strcmp(typeName, callConv.name) == 0)
-            {
-                // Error if there are multiple recognized calling conventions
-                if (found)
-                {
-                    *errorResID = IDS_EE_MULTIPLE_CALLCONV_UNSUPPORTED;
-                    return COR_E_INVALIDPROGRAM;
-                }
-
-                *callConvOut = callConv.value;
-                found = true;
-            }
+            // In this case, the only specified calling convention is CallConvMemberFunction.
+            // Set *callConvOut to the default unmanaged calling convention.
+            *callConvOut = MetaSig::GetDefaultUnmanagedCallingConvention();
         }
+
+        *callConvOut = MetaSig::GetMemberFunctionUnmanagedCallingConventionVariant(*callConvOut);
     }
 
-    return found ? S_OK : S_FALSE;
+    return *callConvOut != CorInfoCallConvExtension::Managed ? S_OK : S_FALSE;
 }
 
+// According to ECMA-335, type name strings are UTF-8. Since we are
+// looking for type names that are equivalent in ASCII and UTF-8,
+// using a const char constant is acceptable. Type name strings are
+// in Fully Qualified form, so we include the ',' delimiter.
+#define MAKE_FULLY_QUALIFIED_CALLCONV_TYPE_NAME_PREFIX(callConvTypeName) CMOD_CALLCONV_NAMESPACE "." callConvTypeName ","
+
+namespace
+{
+    // Templated function to compute if a char string begins with a constant string.
+    template<size_t S2LEN>
+    bool BeginsWith(size_t s1Len, const char* s1, const char (&s2)[S2LEN])
+    {
+        WRAPPER_NO_CONTRACT;
+
+        size_t s2Len = S2LEN - 1; // Remove null
+
+        if (s1Len < s2Len)
+            return false;
+
+        return (0 == strncmp(s1, s2, s2Len));
+    }
+}
+
+bool MetaSig::TryApplyModOptToCallingConvention(
+            LPCSTR callConvModOptName,
+            size_t callConvModOptNameLength,
+            CallConvModOptNameType nameType,
+            CorInfoCallConvExtension* pBaseCallConv,
+            CallingConventionModifiers* pCallConvModifiers)
+{
+#define BASE_CALL_CONV(callConvModOptMetadataName, CallConvExtensionMember)       \
```

Comment:
Not really a fan of this macro magic to be honest. You've done a good job making updating it simple so lets go with it, but note that stepping through this code is basically impossible now.

Reply:
Yeah I don't really like it either but I couldn't think of another way to avoid code duplication. If we're fine with code-duplication, I can inline some or all of the macros depending on how much we want to duplicate.

Reply:
I am okay with it. I may try and rework it at some point but I am fine with it if you are for the time being.

Reply:
Yeah I'm fine with it for now.